ethcore/src/miner/banning_queue.rs

```diff
@@ -78,6 +79,7 @@ impl BanningTransactionQueue {
 	/// May reject transaction because of the banlist.
 	pub fn add_with_banlist<F, G>(
 		&mut self,
+		chain: Option<&MiningBlockChainClient>,
```

Comment:
IMHO it would be better to avoid passing whole `Client` to transaction queue, I was trying to abstract out parts that are needed. Seems that with `ZeroGasPriceChecker` it could also play nicely (you get get rid of couple of `Option` and `Box`)

```rust
pub fn add_with_banlist<F, G, H>(
  &mut self,
  transaction: SignedTransaction,
  time: QueingInstant,
  account_details: &F,
  gas_estimator: &G,
  zgp_checker: &H,
) -> ... where
  F: Fn(&Address) -> AccountDetails,
  G: Fn(&SignedTransaction) -> U256,
  H: Fn(&Address) -> bool/Result<(), String>
{
}
```
In such case you don't need neither `Option<&MiningBlockChainClient>` nor `Option<Box<GasPriceChecker>>`

Since we already have 3 generic functions, probably it could be further refactored to a single trait returning all those details.